types/set.go

```diff
@@ -3,18 +3,83 @@ package types
 import (
 	"bytes"
 	"encoding/json"
+	"slices"
+
+	"golang.org/x/exp/maps"
 )
 
-// A Set is a collection of elements that can be of the same or different types.
-type Set []Value
+// A Set is an immutable collection of elements that can be of the same or different types.
+type Set struct {
+	s       map[uint64]Value
+	hashVal uint64
+}
+
+// NewSet takes a slice of Values and stores a clone of the values internally.
+func NewSet(v []Value) Set {
+	orig := make([]Value, len(v))
+	set := make(map[uint64]Value, len(v))
+
+	for i, vv := range v {
+		orig[i] = vv
+
+		hash := vv.hash()
```

Comment:
Another thought. Could we just call `vv.String()` here and hash the string output of the `Value`? We'd have to check if the `String()` output is always the same for any equivalent instantiation of `Value`...

Just a thought before I step away for the day...

Reply:
I'd vote against this, since creating the full String is so costly.